### PR TITLE
`IntentTEDPolicy`: `rasa interactive` updates

### LINKED ISSUE #8421: IntentTEDPolicy: Adapt `rasa interactive`
**Description of Problem**:
When there is an unlikely intent, `IntentTEDPolicy` predicts an `action_unlikely_intent` to indicate the unlikely intent. This could be confusing to the user as they don't know where this action is coming from.

**Overview of the Solution**:
When the model predicted this action as part of `rasa interactive` we want to give them additional explanation why this action was predicted. Users can then choose to confirm / deny the action prediction as usual but it doesn't really matter as running this action doesn't have any implications.

[See more information in the implementation proposal.](https://www.notion.so/rasa/Implementation-Proposal-IntentTedPolicy-f16c3eb785964b1ca1249d508f6dc234#b491aa7f877b4db7a8629af1029d288c)

**Blockers** (if relevant):
- [x] Implemented `IntentTEDPolicy`
- [x] https://github.com/RasaHQ/rasa/issues/8417

**Definition of Done**:
- [ ] Interactive displays the explanation for the prediction of `action_unlikely_intent` to the user.
- [ ] somebody from CSE double checked the UX of the explanation

### PR DESCRIPTION
**Pull request by alwx**
**Proposed changes**:
- Adapts `rasa interactive` to `IntentTEDPolicy` (`action_unlikely_intent`)
- fixes #8421 

**It requires `rasa test` updates to be merged first! https://github.com/RasaHQ/rasa/pull/8843**

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by alwx** [comment]
There are literally no tests for `rasa interactive` and `_validate_action` function that's why nothing was added here as well. Probably the right action would be to add an issue to add tests for `rasa interactive`.

File: rasa/core/training/interactive.py [hunk 1]
```diff
@@ -1130,9 +1131,22 @@ async def _validate_action(
...
     Returns `True` if the prediction is correct, `False` otherwise."""
 
-    question = questionary.confirm(f"The bot wants to run '{action_name}', correct?")
-
-    is_correct = await _ask_questions(question, conversation_id, endpoint)
+    if action_name == ACTION_UNLIKELY_INTENT_NAME:
+        print(
+            f"The bot wants to run '{action_name}' "
+            f"(https://rasa.com/docs/rasa/default-actions#action_unlikely_intent) "
+            f"to indicate that this user message was unexpected "
```
Comment (ancalita) [inline rasa/core/training/interactive.py:1138]: Should this be rephrased to `the last user message was unexpected`?

File: rasa/core/training/interactive.py [hunk 2]
```diff
@@ -1130,9 +1131,22 @@ async def _validate_action(
...
+            f"given the context of the conversation. "
+            f"Check out IntentTEDPolicy to learn more. "
```
Comment (ancalita) [inline rasa/core/training/interactive.py:1140]: Should we link to the policy docs here?

File: rasa/core/training/interactive.py [hunk 3]
```diff
@@ -1130,9 +1131,22 @@ async def _validate_action(
...
+            f"Press any key to continue…"
+        )
+        input()
+        is_correct = True
```
Comment (ancalita) [inline rasa/core/training/interactive.py:1144]: I think this part of the GH solution is not implemented:`Users can then choose to confirm / deny the action prediction`, is that something that was decided after the issue submission?

File: rasa/core/training/interactive.py [hunk 4]
```diff
@@ -1130,9 +1131,22 @@ async def _validate_action(
 
     Returns `True` if the prediction is correct, `False` otherwise."""
 
-    question = questionary.confirm(f"The bot wants to run '{action_name}', correct?")
-
-    is_correct = await _ask_questions(question, conversation_id, endpoint)
+    if action_name == ACTION_UNLIKELY_INTENT_NAME:
+        print(
+            f"The bot wants to run '{action_name}' "
+            f"(https://rasa.com/docs/rasa/default-actions#action_unlikely_intent) "
```
Comment (wochinge) [inline rasa/core/training/interactive.py:1137]:
```suggestion
            f"{DOCS_URL_DEFAULT_ACTIONS}#action_unlikely_intent) "
```

File: rasa/core/training/interactive.py (hunk 1 above)
Comment (wochinge) [reply rasa/core/training/interactive.py:1138]: I'd also explain what it means that a message is unexpected, e.g. `the last user message was unexpected at this point in the conversation`. How about asking somebody from CSE for review?

File: rasa/core/training/interactive.py [hunk 5]
```diff
@@ -1130,9 +1131,22 @@ async def _validate_action(
...
+            f"to indicate that this user message was unexpected "
+            f"given the context of the conversation. "
+            f"Check out IntentTEDPolicy to learn more. "
+            f"Press any key to continue…"
+        )
+        input()
```
Comment (wochinge) [inline rasa/core/training/interactive.py:1143]: how about using `questionary` here?

**Review by wochinge — COMMENTED** [review]
I know it's a hard to test module but could we please try to add a test for the new edge case?

File: rasa/core/training/interactive.py (hunk 5 above)
Comment (alwx) [reply rasa/core/training/interactive.py:1143]: From the description: "Users can then choose to confirm / deny the action prediction as usual but it doesn't really matter as running this action doesn't have any implications."
@wochinge  So my question is: if the answer doesn't matter, why do we need to ask a question?

File: rasa/core/training/interactive.py (hunk 1 above)
Comment (alwx) [reply rasa/core/training/interactive.py:1138]: The original message was written by @TyDunn so it's more like a question to you.

File: rasa/core/training/interactive.py (hunk 2 above)
Comment (alwx) [reply rasa/core/training/interactive.py:1140]: Added them.

File: rasa/core/training/interactive.py (hunk 3 above)
Comment (alwx) [reply rasa/core/training/interactive.py:1150]: It's related to my question above.

File: rasa/core/training/interactive.py (hunk 1 above)
Comment (TyDunn) [reply rasa/core/training/interactive.py:1144]: `the last user message was unexpected at this point in the conversation` is an improvement. We discussed with CSE here: https://rasa-hq.slack.com/archives/C36SS4N8M/p1623402780458000

File: rasa/core/training/interactive.py (hunk 5 above)
Comment (wochinge) [reply rasa/core/training/interactive.py:1147]: I see your point. I'd asking the question makes sense as
* aligns with behavior for other predicted actions
* makes uses explicitly think about what happened. If it's just a print statement they might just skim over it without actually considering the implications.

**Review by wochinge — COMMENTED** [review]
Can we please add a test for it, Alex?

File: tests/core/training/test_interactive.py [hunk 6]
```diff
@@ -733,6 +737,51 @@ def test_retry_on_error_success(monkeypatch: MonkeyPatch):
...
+    )
+    monkeypatch.setattr(
+        interactive, "_ask_questions", asyncio.coroutine(Mock(return_value=True))
+    )
+    monkeypatch.setattr(interactive, "_form_is_rejected", Mock(return_value=False))
+    monkeypatch.setattr(interactive, "_form_is_restored", Mock(return_value=False))
+    monkeypatch.setattr(interactive, "send_action", asyncio.coroutine(Mock()))
+
+    mocked_confirm = Mock(return_value=None)
+    monkeypatch.setattr(interactive.questionary, "confirm", mocked_confirm)
```
Comment (wochinge) [inline tests/core/training/test_interactive.py:776]: how about testing this with a `True` and `False` value to cover both paths?

File: tests/core/training/test_interactive.py [hunk 7]
```diff
@@ -733,6 +737,51 @@ def test_retry_on_error_success(monkeypatch: MonkeyPatch):
...
+    monkeypatch: MonkeyPatch,
+    mock_endpoint: EndpointConfig,
+    action_name: Text,
+    question: Text,
+):
+    # mock those functions that shouldn't be called
+    monkeypatch.setattr(
+        interactive,
+        "retrieve_tracker",
+        asyncio.coroutine(
```
Comment (wochinge) [inline tests/core/training/test_interactive.py:764]: asyncio.coroutine is deprecated as of Python 3.8 - can we please us `async def` instead?

File: tests/core/training/test_interactive.py [hunk 8]
```diff
@@ -733,6 +737,51 @@ def test_retry_on_error_success(monkeypatch: MonkeyPatch):
...
+            Mock(return_value=DialogueStateTracker.from_events("one", []))
+        ),
+    )
+    monkeypatch.setattr(
+        interactive, "_ask_questions", asyncio.coroutine(Mock(return_value=True))
+    )
+    monkeypatch.setattr(interactive, "_form_is_rejected", Mock(return_value=False))
+    monkeypatch.setattr(interactive, "_form_is_restored", Mock(return_value=False))
```
Comment (wochinge) [inline tests/core/training/test_interactive.py:772]: why do we need to mock these two? If we provide a good mock tracker then this should just give us the desired results, no?

File: tests/core/training/test_interactive.py [hunk 9]
```diff
@@ -733,6 +737,51 @@ def test_retry_on_error_success(monkeypatch: MonkeyPatch):
...
+    monkeypatch.setattr(interactive, "send_action", asyncio.coroutine(Mock()))
```
Comment (wochinge) [inline tests/core/training/test_interactive.py:773]: how about making sure this gets called with the expected values?

Comment (alwx) [reply tests/core/training/test_interactive.py]: Done

File: rasa/core/training/interactive.py [hunk 10]
```diff
@@ -1130,9 +1136,23 @@ async def _validate_action(
...
     Returns `True` if the prediction is correct, `False` otherwise."""
 
-    question = questionary.confirm(f"The bot wants to run '{action_name}', correct?")
+    if action_name == ACTION_UNLIKELY_INTENT_NAME:
+        question = questionary.confirm(
+            f"The bot wants to run '{action_name}' "
+            f"to indicate that the last user message was unexpected "
+            f"at this point in the conversation. "
+            f"Check out IntentTEDPolicy ({DOCS_URL_POLICIES}/#intent-ted-policy) "
+            f"to learn more."
```
Comment (ancalita) [inline rasa/core/training/interactive.py:1145]: Do we need to ask the user here to press Y/N to continue?

Comment (alwx) [reply rasa/core/training/interactive.py:1145]: That's what we've decided to do, yes (to keep it consistent with the rest)

File: rasa/core/training/interactive.py [hunk 11]
```diff
@@ -1130,9 +1136,23 @@ async def _validate_action(
...
+        )
+    else:
+        question = questionary.confirm(
+            f"The bot wants to run '{action_name}', correct?"
+        )
 
-    is_correct = await _ask_questions(question, conversation_id, endpoint)
+    is_correct = (
+        await _ask_questions(question, conversation_id, endpoint)
+        or action_name == ACTION_UNLIKELY_INTENT_NAME
```
Comment (ancalita) [inline rasa/core/training/interactive.py:1154]: Does this line imply that regardless of user answer to the query, the action is always correct for `action_unlikely_intent`?

Comment (alwx) [reply rasa/core/training/interactive.py:1154]: It means that the choice (Y/N) doesn't affect what's gonna happen next.

File: rasa/core/training/interactive.py (hunk 10 above)
Comment (ancalita) [reply rasa/core/training/interactive.py:1145]: Ah ok, then maybe it should be explicitly stated? Unless `questionary` takes care of this 😄

**Review by wochinge — APPROVED** [review]
Thanks for adding the test! 💯

File: rasa/core/training/interactive.py (hunk 10 above)
Comment (alwx) [reply rasa/core/training/interactive.py:1145]: IMO this should instead be written in docs with the explanation what `action_unlikely_intent` is and how it works. We can also write it here but I am afraid the message will be too long in this case.

Comment (ancalita) [reply rasa/core/training/interactive.py:1145]: Ah I meant maybe it's not obvious what keys to press from the question and that should be explicitly stated, thats all 👍

Comment (alwx) [reply rasa/core/training/interactive.py:1145]: Oh, got it! Actually, `questionary` handles that and shows the available options automatically :)
